File: examples/library/borrow_book.go
package library

import (
	"context"
	"errors"

	"github.com/err0r500/fairway"
	"github.com/err0r500/fairway/dcb"
)

// EVENTS

type BookBorrowed struct {
	BookId     string `json:"book_id"`
	BorrowerId string `json:"borrower_id"`
}

func (e BookBorrowed) Tags() []string {
	return []string{"book_id:" + e.BookId, "borrower_id:" + e.BorrowerId}
}

type BookReturned struct {
	BookId     string `json:"book_id"`
	BorrowerId string `json:"borrower_id"`
}

func (e BookReturned) Tags() []string {
	return []string{"book_id:" + e.BookId, "borrower_id:" + e.BorrowerId}
}

// COMMAND

var ErrBookAlreadyBorrowed = errors.New("book is already borrowed")
var ErrBorrowerLimitReached = errors.New("borrower has reached max 5 books")

type BorrowBook struct {
	BookId     string
	BorrowerId string
}

const maxBooksPerBorrower = 5

func (cmd BorrowBook) Run(ctx context.Context, ev fairway.EventReadAppender) error {
	if err := cmd.ensureBookNotCurrentlyBorrowed(ctx, ev); err != nil {
		return err
	}

	if err := cmd.ensureReaderBelowMaxBorrowedBooksLimit(ctx, ev); err != nil {
		return err
	}

	return ev.AppendEvents(ctx, fairway.NewEvent(BookBorrowed{
		BookId:     cmd.BookId,
		BorrowerId: cmd.BorrowerId,
	}))
}

func (cmd BorrowBook) ensureBookNotCurrentlyBorrowed(ctx context.Context, ev fairway.EventsReader) error {
	isBorrowed := false

	if err := ev.ReadEvents(ctx,
		fairway.QueryItems(
			fairway.NewQueryItem().
				Types(BookBorrowed{}, BookReturned{}).
				Tags("book_id:"+cmd.BookId),
		).WithOptions(dcb.ReadOptions{Reverse: true, Limit: 1}),
		func(e fairway.Event) bool {
			_, isBorrowed = e.Data.(BookBorrowed)
			return false // only need first event
		}); err != nil {
		return err
	}

	if isBorrowed {
		return ErrBookAlreadyBorrowed
	}

	return nil
}

func (cmd BorrowBook) ensureReaderBelowMaxBorrowedBooksLimit(ctx context.Context, ev fairway.EventsReader) error {
	borrowerBooks := make(map[string]bool)
	if err := ev.ReadEvents(ctx,
		fairway.QueryItems(
			fairway.NewQueryItem().
				Types(BookBorrowed{}, BookReturned{}).
				Tags("borrower_id:"+cmd.BorrowerId),
		),
		func(e fairway.Event) bool {
			switch data := e.Data.(type) {
			case BookBorrowed:
				borrowerBooks[data.BookId] = true
			case BookReturned:
				delete(borrowerBooks, data.BookId)
			}
			return true
		}); err != nil {
		return err
	}

	if len(borrowerBooks) >= maxBooksPerBorrower {
		return ErrBorrowerLimitReached
	}

	return nil
}

File: fairway_ex/lib/examples/library/borrow_book.ex
defmodule Library.Event.BookBorrowed do
  @moduledoc "A book was borrowed."

  @behaviour Fairway.EventData
  @derive Jason.Encoder
  defstruct [:book_id, :borrower_id]

  @impl true
  def tags(%{book_id: bid, borrower_id: rid}), do: ["book_id:#{bid}", "borrower_id:#{rid}"]

  @impl true
  def type_name, do: "Library.BookBorrowed"
end

defmodule Library.Event.BookReturned do
  @moduledoc "A borrowed book was returned."

  @behaviour Fairway.EventData
  @derive Jason.Encoder
  defstruct [:book_id, :borrower_id]

  @impl true
  def tags(%{book_id: bid, borrower_id: rid}), do: ["book_id:#{bid}", "borrower_id:#{rid}"]

  @impl true
  def type_name, do: "Library.BookReturned"
end

defmodule Library.Change.BorrowBook do
  @moduledoc """
  Pure command: borrow a book from the library.

  Two constraints are enforced via DCB:
  1. The book must not currently be borrowed (checked by looking at the most
     recent book-scoped event — if it's a `BookBorrowed`, the book is out).
  2. The borrower must not have reached the maximum of #{Library.Change.BorrowBook.max_books_per_borrower()}
     currently borrowed books.

  Both checks read different slices of the event log and each contributes
  its own `AppendCondition`, so a conflicting concurrent write to either
  slice causes a clean retry.
  """

  alias Fairway.{Query, QueryItem, ReadAppender, Event}
  alias Library.Event.{BookBorrowed, BookReturned}

  @max_books 5
  def max_books_per_borrower, do: @max_books

  def run(ra, %{book_id: book_id, borrower_id: borrower_id}) do
    with {:ok, ra} <- ensure_book_available(ra, book_id),
         {:ok, ra} <- ensure_borrower_below_limit(ra, borrower_id) do
      ReadAppender.append_events(ra, [
        Event.new(%BookBorrowed{book_id: book_id, borrower_id: borrower_id})
      ])
    end
  end

  # ── Private ──────────────────────────────────────────────────────────────────

  # Read the most recent event for this book (reverse: true, limit: 1).
  # If it's a BookBorrowed the book is still out; if BookReturned (or absent)
  # it is available.
  defp ensure_book_available(ra, book_id) do
    query =
      Query.new(
        [QueryItem.new(types: [BookBorrowed, BookReturned], tags: ["book_id:#{book_id}"])],
        %{reverse: true, limit: 1}
      )

    {ra, is_borrowed} =
      ReadAppender.fold_events(ra, query, false, fn
        %{data: %BookBorrowed{}}, _ -> {false, true}
        %{data: %BookReturned{}}, _ -> {false, false}
        _, acc -> {true, acc}
      end)

    if is_borrowed do
      {:error, :book_already_borrowed}
    else
      {:ok, ra}
    end
  end

  # Count how many books the borrower currently holds (borrowed minus returned).
  defp ensure_borrower_below_limit(ra, borrower_id) do
    query =
      Query.new([
        QueryItem.new(
          types: [BookBorrowed, BookReturned],
          tags: ["borrower_id:#{borrower_id}"]
        )
      ])

    {ra, held} =
      ReadAppender.fold_events(ra, query, MapSet.new(), fn
        %{data: %BookBorrowed{book_id: bid}}, acc -> {true, MapSet.put(acc, bid)}
        %{data: %BookReturned{book_id: bid}}, acc -> {true, MapSet.delete(acc, bid)}
        _, acc -> {true, acc}
      end)

    if MapSet.size(held) >= @max_books do
      {:error, :borrower_limit_reached}
    else
      {:ok, ra}
    end
  end
end
